### PR TITLE
Add Nonnull annotations to StreamSerializer.

### PR DESCRIPTION
**Pull request by aaronriekenberg**

### CODE REVIEW

**Review by kellar — APPROVED** [review]
Change looks good. Though I am not a code owner/admin, I am approving this PR as part of Target's open source contribution policy requiring an approval from someone at Target.

**Comment by sancar** [comment]
run-lab-run

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed.
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal com.github.spotbugs:spotbugs-maven-plugin:4.1.4:check (default) on project hazelcast: failed with 1 bugs and 0 errors -> [Help 1]
--------------------------
</pre>
</details>

**Comment by sancar** [comment]
Hi @aaronriekenberg Thanks for the contribution. 

A small warning, there is one place that we pass `null` to StreamSerializer as far as remember. It is `nullSerializer`.  A `null` is never passed to the user-configured StreamSerializer. 
You may need to add a suppression to static checkers for that case if they complain. 

It seems that spotbugs is complaining about something. 

You can run the following two commands to make sure they are passing in your local before sending.
	mvn clean install -DskipTest  -Pspotbugs 
	mvn clean install -DskipTests -Pcheckstyle

**Comment by aaronriekenberg** [comment]
Hi @sancar thanks for your review and comments.

Building locally I see the error from spotbugs is:

```
[ERROR] High: com.hazelcast.internal.serialization.impl.defaultserializers.ConstantSerializers$NullSerializer.read(ObjectDataInput) may return null, but is declared @Nonnull [com.hazelcast.internal.serialization.impl.defaultserializers.ConstantSerializers$NullSerializer] At ConstantSerializers.java:[line 62] NP_NONNULL_RETURN_VIOLATION
```

Looks like this is happening because NullSerializer.read() returns null: https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/defaultserializers/ConstantSerializers.java#L62

To avoid the error from spotbugs I have changed the return type from `StreamSerializer.read()` to use `@Nullable` annotation.  Verified this passes spotbugs and checkstyle in local.

For my use case the main question is whether the input parameters to `StreamSerializer.read()` and `StreamSerializer.write()` can ever be null.  So I think it is still useful to have these parameters marked as `@Nonnull` in the StreamSerializer interface if possible.  Thanks!

**Comment by sancar** [comment]
Hi @aaronriekenberg , sorry for not paying attention to this. We had a busy week, and it may took some time before we can look into this. 


Let me share my concerns although I could not decide on how to proceed. I will try to get opinions of my team members as well. 

Both `write` and `read` can get and return `null`. The only internal implementation does it is `nullSerializer`. So if we mark this `Nonnull` on both read and write, `nullSerializer` will be exception for that rule.  
We don't expect any user implemented serializer to return `null`. Because if it were `null` when writing, it will be handled by `nullSerializer` and we will call the our users serializer. 

Having said that, this is an expectation of ours from the users. If some users are already returning `null` for some reason that we can not foresee, then potentially their code will be broken. I am saying that because when we put `Nonnul`, we will probably forced to remove `null` checks by the static analyzers and this will break users code. I want to say this is a wrong usage and we should not be concerned with this problem, but I wanted to make the implications obvious. 

In short, I will try to gather some more opinion on this next week at best. Thanks for your patience.

**Comment by sancar** [comment]
Hi @aaronriekenberg 

Lets add NonNull annotation to both read and write for the object. 
This will make our intention clear to our users.

And then, we need to suppress/refactor `nullSerializer` to make static analyzers pass.

**Comment by mmedenjak** [comment]
run-lab-run

**Comment by mmedenjak** [comment]
Hey @aaronriekenberg ! Are you able to address @sancar 's comment above?

**Comment by aaronriekenberg** [comment]
Hi All - based on comments have made the following updates:

1. Put back the @Nonnull annotation on StreamSerializer.read() return value.
2. Added @SuppressFBWarnings("NP_NONNULL_RETURN_VIOLATION") to NullSerializer.read() since this method returns null.

This passes spotbugs and checkstyle locally:

**Comment by mdumandag** [comment]
run-lab-run

**Review by sancar — APPROVED** [review]

**Comment by sancar** [comment]
@aaronriekenberg Thanks again for the contribution. I have merged the pr.
